tests/5.0/target_data/test_target_data_use_device_addr.F90

```diff
@@ -0,0 +1,67 @@
+!===--- test_target_data_use_device_addr.F90 ------------------------------------------===//
+!
+! OpenMP API Version 5.0 Nov 2018
+!
+! This file is a test for the use_device_addr when used with the map clause with
+! target data directive. This test uses a scalar and an array of size N which values
+! are modified on the  device and tested in the host. List items that  appear in a
+! use_device_addr clause have the address of the corresponding object in the device
+! data environment inside the construct. This test also tests that address conversions
+! of use_device_addr clauses will occur as if performed after all variables are mapped
+! according to those map clauses.
+!
+!===-----------------------------------------------------------------------------------===//
+
+#include "ompvv.F90"
+
+PROGRAM test_target_data_use_device_addr
+   USE iso_fortran_env
+   USE ompvv_lib
+   USE omp_lib
+   implicit none
+
+   REAL, POINTER :: dev_ptr
+   INTEGER :: errors, host_data, device_data
+
+   errors = 0
+   device_data = 14
+   host_data = 0 
+
+   ALLOCATE(dev_ptr)
+
+   OMPVV_TEST_OFFLOADING
+
+   OMPVV_TEST_VERBOSE(test(dev_ptr) .ne. 0)
+
+   OMPVV_REPORT_AND_RETURN()
+
+
+CONTAINS
+   INTEGER FUNCTION test(x)
+      REAL, INTENT(INOUT) :: x
+
+      !$omp target data map(to: device_data)
+
+         !$omp target data use_device_addr(device_data)
+            x = device_data
+         !$omp end target data
+
+         !$omp target map(to:device_data) map(tofrom: errors) map(from:host_data) is_device_ptr(x)
```

Comment:
With this `is_device_ptr` I got completely stuck when studying OpenMP 5.0 – as it is completely unclear what `is_device_ptr` is supposed to mean and what is permitted and not also does not make any sense for Fortran.

OpenMP 5.1 solved all my issues and has:

> The `is_device_ptr`clause indicates that its list items are device pointers if they are of type `C_PTR`. [...] For any list item in an `is_device_ptr`clause that is not of type `C_PTR`, the behavior is as if the list item appeared in a has_device_addr clause.

However, even with this reading it does not make sense. `x` is called from host context and is not mapped to the device (i.e. there is no "corresponding storage in the device data environment"). It's value it taken from the bogus `x = device_data` assignment – but that does not turn it into a "corresponding storage in the device data environment".

Reply:
Thank you for the feedback! I Will take a look at the test-case you referenced above and fix this.